d2/src/main/java/com/linkedin/d2/balancer/util/WarmUpLoadBalancer.java

```diff
@@ -104,105 +107,160 @@ public WarmUpLoadBalancer(LoadBalancerWithFacilities balancer, WarmUpService ser
     _outstandingRequests = new ConcurrentLinkedDeque<>();
     _usedServices = new HashSet<>();
     _dualReadStateManager = dualReadStateManager;
+    _isIndis = isIndis;
+    _printName = String.format("%s WarmUp", _isIndis ? "xDS" : "ZK");
   }
 
   @Override
-  public void start(Callback<None> callback)
-  {
-    LOG.info("D2 WarmUp enabled");
-    _loadBalancer.start(new Callback<None>()
-    {
+  public void start(Callback<None> callback) {
+    LOG.info("{} enabled", _printName);
+    _loadBalancer.start(new Callback<None>() {
       @Override
-      public void onError(Throwable e)
-      {
+      public void onError(Throwable e) {
         callback.onError(e);
       }
 
       @Override
-      public void onSuccess(None result)
+      public void onSuccess(None result) {
+        _allStartTime = SystemClock.instance().currentTimeMillis();
+
+        Future<?> prepareTaskFuture = _executorService.submit(() -> prepareWarmUp());
+        try
+        {
+          prepareTaskFuture.get(_warmUpTimeoutSeconds, TimeUnit.SECONDS);
+        }
+        catch (TimeoutException e)
+        {
+          LOG.info("{} hit timeout: {}s. The WarmUp will continue in background", _printName, _warmUpTimeoutSeconds);
+          callback.onSuccess(None.none());
+        }
```

Comment:
While this is a great idea, it'll end up blocking the caller of the callback (which could be just about any thread? not sure). This is a concern considering the variety of load-balancers this class is used to wrap with.

The root-cause for the timeout not triggering seems to be the scheduled executor behavior of executing all tasks in order, including the scheduled ones. This clearly doesn't work here. So what we might need to do here is make the tasks themselves timeout aware (as in each task checks for whether it has exceeded the warmup timeout) and have them trigger the callback. This can be discussed further.

Reply:
good point. Although only restli client uses WarmUpLoadBalancer (grpc client doesn't), it's not ideal to block the thread that calls the start callback, especially to the property event bus thread that updates the simple load balancer. I've put the work to a timeout-aware, non-blocking task.

Reply:
The solution to this problem looks good.